source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -18,17 +21,117 @@ MapdataCheckPoint::MapdataCheckPoint(const SData *data)
+
+    // Calculate the quadrilateral's `m_prevPoints`. If the check point is the first in its group,
+    // it has multiple previous checkpoints defined by its preceding checkpaths
+    if (m_prevPt == 0xFF) {
+        MapdataCheckPath *checkpath = checkPathAccessor->findCheckpathForCheckpoint(id);
+        if (checkpath) {
+            m_prevCount = 0;
+
+            for (auto [i, prevID] : std::views::enumerate(checkpath->prev())) {
+                if (prevID == 0xFF) {
+                    continue;
+                }
+
+                m_prevPoints[i] = accessor.get(checkPathAccessor->get(prevID)->end());
+                ++m_prevCount;
+            }
+        }
+    } else {
+        m_prevPoints[0] = accessor.get(m_prevPt);
+        ++m_prevCount;
+    }
+
+    // Calculate the quadrilateral's `m_nextPoints`. If the checkpoint is the last in its group, it
+    // can have multiple quadrilaterals (and nextCheckpoint) which are determined by its next
+    // path(s)
+    if (m_nextPt == 0xFF) {
+        MapdataCheckPath *checkpath = checkPathAccessor->findCheckpathForCheckpoint(id);
+        if (checkpath) {
+            m_nextCount = 0;
+
+            for (auto [i, nextID] : std::views::enumerate(checkpath->next())) {
+                if (nextID == 0xFF) {
+                    continue;
+                }
+
+                m_nextPoints[i].checkpoint = accessor.get(checkPathAccessor->get(nextID)->start());
+                ++m_nextCount;
+            }
+        }
+    } else {
+        m_nextPoints[0].checkpoint = accessor.get(m_nextPt);
+        ++m_nextCount;
+    }
+
+    // Form the checkpoint's quadrilateral(s)
+    for (auto [i, next] : std::views::enumerate(m_nextPoints)) {
+        if (i < m_nextCount) {
+            auto &nextLinked = m_nextPoints[i];
+            auto *nextPoint = nextLinked.checkpoint;
+
+            next.distance = (nextPoint->m_midpoint - m_midpoint).normalise();
+            next.p0diff = nextPoint->m_left - m_left;
+            next.p1diff = nextPoint->m_right - m_right;
+        } else {
+            next.distance = 0.0f;
+            next.p0diff = EGG::Vector2f::zero;
+            next.p1diff = EGG::Vector2f::zero;
+        }
+    }
+}
+
+s8 MapdataCheckPoint::checkArea() const {
+    return m_checkArea;
+}
+
+MapdataCheckPoint *MapdataCheckPoint::nextPoint(size_t i) const {
+    ASSERT(i < m_nextPoints.size());
+    return m_nextPoints[i].checkpoint;
+}
+
+/// @addr{0x80515244}
+/// @brief Initializes all checkpoint links, and finds the finish line and last key checkpoint.
+void MapdataCheckPointAccessor::init() {
+    s8 lastKcpType = -1;
+    s16 finishLineCheckpointId = -1;
+
+    for (size_t ckptId = 0; ckptId < size(); ckptId++) {
+        MapdataCheckPoint *checkpoint = get(ckptId);
+        s8 checkArea = checkpoint->checkArea();
+        checkpoint->initCheckpointLinks(*this, ckptId);
+
+        if (checkArea == static_cast<s8>(MapdataCheckPoint::CheckArea::FINISH_LINE)) {
```

Comment:
```suggestion
        if (static_cast<std::underlying_type_t<MapdataCheckPoint::CheckArea>>(checkArea) == MapdataCheckPoint::CheckArea::FINISH_LINE) {
```
This is fine to split out into a helper function if this suggestion looks too ugly - `isCheckAreaFinishLine`.

Reply:
I get: Invalid operands to binary expression ('std::underlying_type_t<MapdataCheckPoint::CheckArea>' (aka 'int') and 'System::MapdataCheckPoint::CheckArea')

Reply:
it likes `if (static_cast<MapdataCheckPoint::CheckArea>(checkArea) == MapdataCheckPoint::CheckArea::FINISH_LINE) {`

Reply:
Sorry, you're right. Yeah, just static cast to the enum.